CHANGELOG.md

```diff
@@ -93,7 +93,9 @@ Ref: https://keepachangelog.com/en/1.0.0/
 
 ### CLI Breaking Changes
 
-* /
+* (client) [#12723](https://github.com/cosmos/cosmos-sdk/pull/12723) The client config now includes default values for several gas-related parameters, which previously needed to be provided by `--gas`, `--gas-adjustment`, and `--gas-prices` flags.
```

Comment:
Curious, why do you consider this CLI breaking and not a feature or improvement?

Reply:
because users need to add the three lines to their `.simapp/config/client.toml`:

```toml
gas = "200000"
gas-adjustment = "1"
gas-prices = ""
```

either by running `simd init` again or by manually editing the file, of the `tx` command won't work.

this is because when creating the `ClientCtx` the client will attempt to read these three params from `client.toml`, and if not found it will throw an error.

Reply:
Can you add exactly that in the upgrading.md ? Under a Configuration section or something?

Reply:
> Can you add exactly that in the upgrading.md ? Under a Configuration section or something?

done: https://github.com/cosmos/cosmos-sdk/pull/12723/commits/7c099f81b93c27f2750da6dc6ea2cef094669b37

Reply:
Noooo, I don't think this is wise. We should not be requiring _any_ fields in `client.toml` whatsoever. I'm strongly opposed to this. `client.toml` should be completely opt-in and read only when it exists.

Reply:
> Noooo, I don't think this is wise. We should not be requiring _any_ fields in `client.toml` whatsoever. I'm strongly opposed to this. `client.toml` should be completely opt-in and read only when it exists.

this is in fact already the case for other fields currently in `client.toml`. the chain id for example, if you neither have  it configured in `client.toml` nor specify it by the CLI flag, then the client won't be able to sign txs:

https://github.com/cosmos/cosmos-sdk/blob/dc95e33efeb1211b1121f4bbaa3a2b37945ee553/client/tx/factory.go#L251-L253

the gas parameters aren't different from this. you need to either configure them in `client.toml`, or specify with CLI flags.

Reply:
@alexanderbez ok here's how i think we can make this work. currently, `clientCtx` is created by parsing `client.toml`:

```go
// cosmos-sdk/client/config/config.go
func ReadFromClientConfig(ctx client.Context) (client.Context, err) {
	// parse gas setting
	gasSetting, err := client.ParseGasSetting(conf.Gas)
	if err != nil {
		return ctx, fmt.Errorf("couldn't get gas setting from client config: %v", err)
	}

	ctx = ctx.WithGasSetting(gasSetting)

	// parse gas adjustment
	gasAdj, err := strconv.ParseFloat(conf.GasAdjustment, 64)
	if err != nil {
		gasAdj, _ = strconv.ParseFloat(gasAdjustment, 64)
	}

	ctx = ctx.WithGasAdjustment(gasAdj)

	// parse gas prices
	gasPrices, err := sdk.ParseDecCoins(conf.GasPrices)
	if err != nil {
		return ctx, fmt.Errorf("couldn't get gas prices from client config: %v", err)
	}

	ctx = ctx.WithGasPrices(gasPrices)

	// parse other client.toml fields, e.g. chain id, rpc node, sign mode, etc.
	// ...

	return ctx
}
```

take `gas-adjustment` as an example: assume it is NOT present in `client.toml` (e.g. if `client.toml` was generated by an old version of SimApp). in this case `conf.GasAdjustment` will be an empty string `""`, which causes `strconv.ParseFloat` to fail.

this is the reason that the three gas parameters are mandatory in `client.toml`.

however we can make them optional simply by, instead of returning `err` here, we use some default values. this way, users won't need to update their `client.toml` files. sounds good?

Reply:
> In general the order of precedence is:
> 
> env var > file > flag > default

let me research how other config params are handled. i'll keep this PR as draft for now.

Reply:
> however we can make them optional simply by, instead of returning err here, we use some default values. this way, users won't need to update their client.toml files. sounds good?

Yes, in general this is the idea. I would do whatever we do for all the values we support in `client.toml`. I haven't looked at how that file is parsed and handled, but I imagine it only parses when non-empty.